app/src/org/commcare/activities/FormEntryActivity.java

```diff
@@ -707,10 +712,10 @@ private void saveDataToDisk(boolean exit, boolean complete, String updatedSaveNa
             }
             return;
         }
-        // save current answer; if headless, don't evaluate the constraints
-        // before doing so.
+        // Save current answer; if headless, don't evaluate the constraints before doing so.
+        boolean evaluateConstraints = !headless;
         boolean wasScreenSaved =
-                saveAnswersForCurrentScreen(FormEntryConstants.DO_NOT_EVALUATE_CONSTRAINTS, complete, headless);
+                saveAnswersForCurrentScreen(evaluateConstraints, complete, headless);
```

Comment:
IMPORTANT: This change is separate from the rest of the PR, and I'm not sure whether it's right or not, although it seems like it should be... Previously, it looks like we were never evaluating constraints on `saveAnswersForCurrentScreen()` when actually saving the data to disk. That seemed odd to me, and then when reading the comment, also seemed contrary to the sentiment expressed there. So I changed this to reflect the comment instead. If anyone sees a potential issue with this, please flag.